### ISSUE TITLE
Workspace and query windows not saved on restart (RM #5193)

### ISSUE THREAD

**Issue by dpage** [pgadmin-org/pgadmin4#3319]
***Issue migrated from Redmine: https://redmine.postgresql.org/issues/5193***
*Originally created by **Bilge B** at 2020-02-15 09:36:55 UTC.*

If the application is restarted, either deliberately or otherwise, all the work in however many open query windows is permanently lost. It should be clear that preserving work and the application state is better than not, whether or not the user actually cares about the content of those windows, but especially when they do. I realize each query can be manually saved to disk, but even in that case, the same open query windows should be presented when the application reloads so the workspace state is always preserved instead of reset to a blank slate each time.

**Comment by ajsmith-savills** [comment]
This would be an extremely helpful feature, is it still slated to be developed?

**Comment by Bilge** [comment]
This feature has been requested every day for the past decade. They are never going to do it.

**Comment by anthonydb** [comment]
I see I filed a duplicate request for this. I hope you'll prioritize it.

**Comment by Bilge** [comment]
They will never.

**Comment by Amnesiac9** [comment]
This would be really nice! A definite missing feature coming from mysql workbench.

**Comment by yogeshmahajan-1903** [comment]
As part of the feature we have added new preference 'Miscellaneous > User Interface > Save Application State',  whether to preserve tools windows and data. If this setting is turned on, tools window and its data will be saved to pgadmin configuration db. On abrupt shutdown/ refresh of the pgadmin , these tools will be restored with the the respective data.

**Comment by yogeshmahajan-1903** [comment]
@Bilge , @anthonydb 
May you can give a try on [snapshot](https://www.postgresql.org/ftp/pgadmin/pgadmin4/snapshots/) build before it release .

**Comment by aergul** [comment]
@yogeshmahajan-1903 
I installed the latest snapshot.
I noted that `Save the application state?` was set to true by default.
I opened the `CREATE SCRIPT` for a table as well as a query window in which I wrote a query and executed it.
I then force quit pgadmin.
Upon running pgadmin again, the two query windows were restored with the SQL they contained previously.  Awesome!

The window where I had run the query and had the resulting data table did not preserve the data table but I'm not sure if if that's what's meant by:
> tools window and its data will be saved

I noted that the `Object Explorer` was not restored to prior state. Would make a nice stretch goal.

Running on M4 Mac Sequoia 15.1.1

I am glad this feature finally saw the light of the day. Thank you!

**Comment by yogeshmahajan-1903** [comment]
> [@yogeshmahajan-1903](https://github.com/yogeshmahajan-1903) I installed the latest snapshot. I noted that `Save the application state?` was set to true by default. I opened the `CREATE SCRIPT` for a table as well as a query window in which I wrote a query and executed it. I then force quit pgadmin. Upon running pgadmin again, the two query windows were restored with the SQL they contained previously. Awesome!
> 
> The window where I had run the query and had the resulting data table did not preserve the data table but I'm not sure if if that's what's meant by:
> 
> > tools window and its data will be saved
> 
> I noted that the `Object Explorer` was not restored to prior state. Would make a nice stretch goal.
> 
> Running on M4 Mac Sequoia 15.1.1
> 
> I am glad this feature finally saw the light of the day. Thank you!

You are correct, result grid data will NOT be restored.It will only restore only query editor content. 
In case of 
Schema diff  - Only selected servers/db/schemas (not a comparison result), 
Psql - Only the psql windows with connection (no queries will be preserved.)

**Comment by Bilge** [comment]
Looks good to me.

* ✅ Option enabled by default.
* ✅ Restores my query windows with the queries intact.

That's all I ever wanted! Is this data still going to be persisted when upgrading?

**Comment by anthonydb** [comment]
Opened three files and closed the program. All three were in the workspace when I re-opened it, so seems to work. Nice feature, thank you for implementing.

**Comment by pravesh-sharma** [comment]
Found the 2 issues below during initial testing:

1. If I have multiple tools tabs open and then continuously close them, and then close and reopen pgAdmin, only one tab is closed while the others remain open (If I have different tools tab opened and close them then also behaviour is not consistent).
2. The schema diff tool state is not being saved.

**Comment by Bilge** [comment]
⚠ One issue I have found is that, although my query windows are saved, their order is not. In fact, when I restarted my computer and re-opened the application, it seemed my tabs were loaded in reverse order... very confusing!

**Comment by Bilge** [comment]
❗ OK this is definitely _not_ working properly. I keep closing tabs, but every time I restart the application, more and more dead tabs that I closed days or even weeks ago keep reappearing, even tabs that I have now closed multiple times keep being resurrected by restarting the app.

**Comment by yogeshmahajan-1903** [comment]
> ❗ OK this is definitely _not_ working properly. I keep closing tabs, but every time I restart the application, more and more dead tabs that I closed days or even weeks ago keep reappearing, even tabs that I have now closed multiple times keep being resurrected by restarting the app.

Above issue was reported while testing. Fix is available in 9.5 release or you can try the latest nightly build.

**Comment by pravesh-sharma** [comment]
Tested and verified on latest candidate and snapshot  build.

Created a separate issue for tab order not restoring correctly. #8901

**Comment by Bilge** [comment]
I just "up"graded to 9.6 and **lost** all my queries! I try to downgrade to 9.5 to get them back, but the installer just blocks me with "A newer version of pgAdmin is already installed" 🤦🏻‍♂️
